Let identifier_friendly cope with unicode in identifiers

PeterJCLaw opened the pull request:
This fixes rendering of /projects/$proj/versions/ when a branch name contains a character which cannot be expressed in ascii.

The test here is very low-level compared the primary symptom; should I be adding something higher level as well?

PeterJCLaw commented:
@berkerpeksag I'm happy to add those extra tests (done in 0b2ed115bd699581dc8e70a67079457c6313c7e2), though I'm not sure what we gain by testing the case of non-unicode strings since (as fat as I know) Django's `CharField`s always return `unicode` instances?

berkerpeksag commented:
You're right. I think adding a test that covers the `if` branch would be enough:

```py
version = new(Version,
              identifier=u'3d92b728b7d7b842259ac2020c2fa389f13aff0d',
              slug=STABLE, verbose_name=STABLE, type=TAG)
# version.identifier_friendly should return u'3d92b728' since re_sha matches
self.assertEqual(version.identifier_friendly, u'3d92b728')
```

PeterJCLaw commented:
Update: I've remove the redundant tests (the ones I added in 0b2ed11) and added a test for the commit hash shortening behaviour as suggested by @berkerpeksag.

berkerpeksag commented on readthedocs/projects/version_handling.py:
```diff
@@ -108,7 +108,7 @@ def version_windows(versions, major=1, minor=1, point=1):
     try:
         return Version(version_string)
-    except InvalidVersion:
+    except (InvalidVersion, UnicodeEncodeError):
```
Do we need to catch `UnicodeEncodeError` anymore? We already fixed the main reason of getting `UnicodeEncodeError` by removing the `str()` call in `Version.identifier_friendly`. I didn't get any test failures when I removed `UnicodeEncodeError`.

PeterJCLaw replied:
Oops! Sorry, this change isn't meant to be here. It's a workaround for https://github.com/pypa/packaging/issues/90 which I had in place but didn't intend to commit. I'll remove it.

PeterJCLaw replied:
https://github.com/rtfd/readthedocs.org/pull/2560/commits/a64d613bcc96aac59131a0d781386e166cd81121

berkerpeksag commented on readthedocs/rtd_tests/tests/test_version_commit_name.py:
```diff
@@ -13,6 +13,19 @@
 class VersionCommitNameTests(TestCase):
+    def test_branch_name_unicode_non_ascii(self):
+        unicode_name = 'abc_\xd1\x84_\xe2\x99\x98'.decode('utf-8')
+
```
Style nit: new line can be deleted

PeterJCLaw replied:
https://github.com/rtfd/readthedocs.org/pull/2560/commits/bdced713e4ca729b336d0957fdf0ba8693036b29

berkerpeksag commented:
I just left two minor comments, but this looks good to me. Thanks!

PeterJCLaw commented:
@berkerpeksag thanks for the review; I believe I've resolved the highlighted issues.

berkerpeksag commented:
Thanks! :)
